File: 338.比特位计数.swift
/*
 * @lc app=leetcode.cn id=338 lang=swift
 *
 * [338] 比特位计数
 */

// @lc code=start
class Solution {
    func countBits(_ n: Int) -> [Int] {
        var res = [Int](repeating: 0, count: n + 1)
        for idx in 0...n {
            res[idx] = countBit(idx)
        }
        return res
    }

    func countBit(_ n: Int) -> Int {
        var res = 0
        var x = n
        while x > 0 {
            x = x & (x - 1)
            res += 1
        }
        return res
    }
}
// @lc code=end

File: 338.比特位计数.cpp
/*
 * @lc app=leetcode.cn id=338 lang=cpp
 *
 * [338] 比特位计数
 */

// @lc code=start
class Solution {
public:
    vector<int> countBits(int n) {
        vector<int> res(n+1, 0);
        for (int i = 0; i <= n; i++) {
            res[i] = countBit(i);
        }
        return res;
    }

    int countBit(int n) {
        int res = 0;
        while(n > 0) {
            res++;
            n = n & (n - 1);
        }
        return res;
    }
};
// @lc code=end

